File: night_01/golang/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Input struct {
	LeftList    []int
	RightList   []int
	RightCounts map[int]int
}

func getInput() (*Input, error) {
	input := &Input{
		RightCounts: make(map[int]int),
	}

	// Open input file
	inputFile, err := os.Open("../data/input.txt")
	if err != nil {
		return nil, err
	}

	// Read input file
	inputScanner := bufio.NewScanner(inputFile)
	for inputScanner.Scan() {
		inputLine := inputScanner.Text()

		// Skip empty lines
		if inputLine == "" {
			continue
		}

		// Parse input line
		inputNums := strings.Fields(inputLine)
		if len(inputNums) != 2 {
			return nil, errors.New("invalid input")
		}

		leftNum, err := strconv.Atoi(inputNums[0])
		if err != nil {
			return nil, err
		}
		input.LeftList = append(input.LeftList, leftNum)

		rightNum, err := strconv.Atoi(inputNums[1])
		if err != nil {
			return nil, err
		}
		input.RightList = append(input.RightList, rightNum)
		input.RightCounts[rightNum]++

	}

	sort.Ints(input.LeftList)
	sort.Ints(input.RightList)

	return input, nil
}

func absDifference(a, b int) int {
	if a > b {
		return a - b
	}
	return b - a
}

func getDistance(input *Input) int {
	var total int
	for i := 0; i < len(input.LeftList); i++ {
		total += absDifference(input.LeftList[i], input.RightList[i])
	}
	return total
}

func getSimilarity(input *Input) int {
	var total int
	for i := 0; i < len(input.LeftList); i++ {
		leftNum := input.LeftList[i]
		total += input.RightCounts[leftNum] * leftNum
	}
	return total
}

func main() {
	input, err := getInput()
	if err != nil {
		log.Fatal(err)
	}
	distance := getDistance(input)
	fmt.Printf("Distance: %d\n", distance)
	similarity := getSimilarity(input)
	fmt.Printf("Similarity: %d\n", similarity)
}

File: night_01/rust/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

#[derive(Debug)]
struct Input {
    left_list: Vec<i32>,
    right_list: Vec<i32>,
    right_counts: HashMap<i32, i32>,
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn get_input() -> Input {
    let mut left_list: Vec<i32> = Vec::new();
    let mut right_list: Vec<i32> = Vec::new();
    let mut right_counts: HashMap<i32, i32> = HashMap::new();

    if let Ok(lines) = read_lines("../data/input.txt") {
        for line in lines {
            if let Ok(ip) = line {
                let parts: Vec<&str> = ip.split_whitespace().collect();
                let left: i32 = parts[0].parse::<i32>().unwrap();
                let right: i32 = parts[1].parse::<i32>().unwrap();
                left_list.push(left);
                right_list.push(right);
                let count = right_counts.entry(right).or_insert(0);
                *count += 1;
            }
        }
    }

    left_list.sort();
    right_list.sort();

    Input {
        left_list,
        right_list,
        right_counts,
    }
}

fn abs_difference(a: i32, b: i32) -> i32 {
    if a > b {
        a - b
    } else {
        b - a
    }
}

fn get_distance(input : &Input) -> i32 {
    let mut distance = 0;
    for i in 0..input.left_list.len() {
        let left = input.left_list[i];
        let right = input.right_list[i];
        distance += abs_difference(left, right);
    }
    distance
}

fn get_similarity(input : &Input) -> i32 {
    let mut similarity = 0;
    for i in 0..input.left_list.len() {
        let left = input.left_list[i];
        let right_count = input.right_counts.get(&left).unwrap_or(&0);
        similarity += left * *right_count;
    }
    similarity
}

fn main() {
    let input = get_input();
    let distance = get_distance(&input);
    let similarity = get_similarity(&input);
    println!("Distance: {:?}", distance);
    println!("Similarity: {:?}", similarity);
}
